hare/eligibility/oracle.go

```diff
@@ -166,69 +166,166 @@ func (o *Oracle) buildVRFMessage(layer types.LayerID, round int32) ([]byte, erro
+	actives, err := o.actives(layer)
+	if err != nil {
+		if err == errGenesis { // we are in genesis
+			return o.genesisTotalWeight / 50, nil
+		}
+
+		o.With().Error("minerWeight erred while calling actives func", log.Err(err), layer)
+		return 0, err
+	}
+
+	w, ok := actives[id.Key]
+	if !ok {
+		o.With().Error("miner is not active in specified layer",
+			log.Int("active_set_size", len(actives)),
+			log.String("actives", fmt.Sprintf("%v", actives)),
+			layer, log.String("id.Key", id.Key),
+		)
+		return 0, errors.New("miner is not active in specified layer")
+	}
+	return w, nil
+}
+
+func calcVrfFrac(sig []byte) fixed.Fixed {
+	sha := sha256.Sum256(sig)
+	return fixed.FracFromBytes(sha[:8])
+}
+
+func (o *Oracle) prepareEligibilityCheck(layer types.LayerID, round int32, committeeSize int, id types.NodeID, sig []byte) (fixed.Fixed, fixed.Fixed, fixed.Fixed, bool, error) {
 	msg, err := o.buildVRFMessage(layer, round)
 	if err != nil {
 		o.Error("eligibility: could not build VRF message")
-		return false, err
+		return fixed.Fixed{}, fixed.Fixed{}, fixed.Fixed{}, true, err
 	}
 
 	// validate message
 	res, err := o.vrfVerifier(msg, sig, id.VRFPublicKey)
 	if err != nil {
 		o.Error("eligibility: VRF verification failed: %v", err)
-		return false, err
+		return fixed.Fixed{}, fixed.Fixed{}, fixed.Fixed{}, true, err
 	}
 	if !res {
 		o.With().Info("eligibility: a node did not pass VRF signature verification",
 			id,
 			layer)
-		return false, nil
+		return fixed.Fixed{}, fixed.Fixed{}, fixed.Fixed{}, true, nil
 	}
 
 	// get active set size
-	activeSetSize, err := o.activeSetSize(layer)
+	totalWeight, err := o.totalWeight(layer)
 	if err != nil {
-		return false, err
+		return fixed.Fixed{}, fixed.Fixed{}, fixed.Fixed{}, true, err
 	}
 
-	// require activeSetSize > 0
-	if activeSetSize == 0 {
-		o.Warning("eligibility: active set size is zero")
-		return false, errors.New("active set size is zero")
+	// require totalWeight > 0
+	if totalWeight == 0 {
+		o.Warning("eligibility: total weight is zero")
+		return fixed.Fixed{}, fixed.Fixed{}, fixed.Fixed{}, true, errors.New("total weight is zero")
 	}
 
 	// calc hash & check threshold
-	sha := sha256.Sum256(sig)
-	shaUint32 := binary.LittleEndian.Uint32(sha[:4])
-	// avoid division (no floating point) & do operations on uint64 to avoid overflow
-	if uint64(activeSetSize)*uint64(shaUint32) > uint64(committeeSize)*uint64(math.MaxUint32) {
-		o.With().Info("eligibility: node did not pass VRF eligibility threshold",
-			id,
-			log.Int("committee_size", committeeSize),
-			log.Uint32("active_set_size", activeSetSize),
-			log.Int32("round", round),
-			layer)
-		return false, nil
+	minerWeight, err := o.minerWeight(layer, id)
+	if err != nil {
+		return fixed.Fixed{}, fixed.Fixed{}, fixed.Fixed{}, true, err
+	}
+
+	// ensure miner weight fits in int
+	if uint64(int(minerWeight)) != minerWeight {
+		panic(fmt.Sprintf("minerWeight overflows int (%d)", minerWeight))
```

Comment:
Looks like it would be better to panic when `minerWeight` was calculated. 
But regardless of that, what's the greater solution besides panicking? 
Is there any other solution besides defining the max allowed bits (64 or more) and define/validate max allowed space+ticks combination and number of miners (for the total calc)?